js/compiler_tests.ts

```diff
@@ -0,0 +1,313 @@
+  originals._os = (compilerInstance as any)._os;
+  originals._ts = (compilerInstance as any)._ts;
+  originals._service = (compilerInstance as any)._service;
+  originals._window = (compilerInstance as any)._window;
+  Object.assign(compilerInstance, mocks);
+});
+
+test(function compilerMakeDefine() {
+  const moduleMetaData = new ModuleMetaData(
+    "/root/project/foo/bar.ts",
+    fooBarTsSource,
+    fooBarTsOutput
+  );
+  const localDefine = compilerInstance.makeDefine(moduleMetaData);
+  let factoryCalled = false;
+  localDefine(
+    ["require", "exports", "compiler"],
+    (_require, _exports, _compiler): void => {
+      factoryCalled = true;
+      assert(
+        typeof _require === "function",
+        "localRequire should be a function"
+      );
+      assert(_exports != null);
+      assert(
+        Object.keys(_exports).length === 0,
+        "exports should have no properties"
+      );
+      assert(compiler === _compiler, "compiler should be passed to factory");
+    }
+  );
+  assert(factoryCalled, "Factory expected to be called");
+});
+
+test(function compilerRun() {
+  const result = compilerInstance.run("foo/bar.ts", "/root/project");
+  assert(result != null);
+});
+
+// TODO testMakeDefineExternalModule - testing that make define properly runs
+// external modules
+
+// TypeScript LanguageServiceHost APIs
+
+test(function compilerGetCompilationSettings() {
+  const result = compilerInstance.getCompilationSettings();
+  for (const key of [
+    "allowJs",
+    "module",
+    "outDir",
+    "inlineSourceMap",
+    "inlineSources",
+    "stripComments",
+    "target"
+  ]) {
+    assert(key in result, `Expected "${key}" in compiler options.`);
+  }
+  assert(Object.keys(result).length === 7, "Expected only 7 keys.");
+});
+
+test(function compilerGetNewLine() {
+  const result = compilerInstance.getNewLine();
+  assert(result === "\n", "Expected newline value of '\\n'.");
+});
+
+test(function compilerGetScriptFileNames() {
+  // Because the compiler contains quite a bit of hidden state, and this
+  // actually is running within an environment with the live compiler
+  // modifying that hidden state is actually quite dangerous, so we are
+  // only going to ensure that that hidden state reflects what will be
+  // communicated to the TypeScript compiler.
+  const modulesLength: number = (compilerInstance as any)._moduleMetaDataMap
+    .size;
+  const result = compilerInstance.getScriptFileNames();
+  assert(result.length !== 0, "Expected result to have length.");
+  assert(result.length === modulesLength, "Expected length does not match.");
+});
+
+test(function compilerGetScriptKind() {
+  assert(compilerInstance.getScriptKind("foo.ts") === 3);
+  assert(compilerInstance.getScriptKind("foo.d.ts") === 3);
+  assert(compilerInstance.getScriptKind("foo.js") === 1);
+  assert(compilerInstance.getScriptKind("foo.json") === 6);
+  assert(compilerInstance.getScriptKind("foo.txt") === 1);
```

Comment:
Why not use `compiler.ScriptKind.JS` and friends ?

Reply:
So there is an issue, for some reason, they don't exist at runtime.  I had to create them in `compiler.ts`, I guess I need to pull them from there too.